feat: LEAP-1424: Open preview window for images in Grid view

hlomzik opened the pull request:
Open a simple modal with image from the data of clicked task. Allow to navigate between tasks visible in Grid view by Prev/Next buttons. Only one image is displayed right now.

Simple common hotkeys are available:
- `<Arrows left/right>` to switch between tasks
- `<Space>` to select/unselect task
- `<Escape>` to close the modal

<img width="1248" alt="Screenshot 2024-12-07 at 19 02 19" src="https://github.com/user-attachments/assets/1ce85919-dbae-4f83-80a4-d175afba5342">


### PR fulfills these requirements
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

### Change has impacts in these area(s)
- [x] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

### Feature flags used to cover this change
`fflag_feat_front_leap_1424_grid_preview_short`

### Describe the reason for change
It's hard to quickly preview task and select it depending on a content in current state — if you open task it's another view and it takes time to load it. Here is the new view for quick preview.

### Does this PR introduce a breaking change?
- [x] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)

Click on task in Grid view now has a different action, but task selection is still possible if you click on task id with checkbox.

### What level of testing was included in the change?
- [ ] e2e
- [ ] integration
- [ ] unit

hlomzik committed:
feat: LEAP-1424: Open preview window for images in Grid view

Open a simple modal with image from the data of clicked task.
Allow to navigate between tasks visible in Grid view by Prev/Next buttons.
Only one image is displayed right now.

hlomzik committed:
Fix missing react hooks imports

hlomzik committed:
Add hotkeys for navigation between pages

hlomzik committed:
Small comment about key and image flickering

hlomzik committed:
Add more hotkeys, improve appearance, add tooltip

Modal is `bare` now to allow to make better styles for it.
UI is more dense and different elements are aligned.

hlomzik committed:
Add FF to control preview in Grid View

hlomzik committed:
Show preview modal only if there is an image in a task

Check fields for "Image" type and show modal only if it exists.

hlomzik committed:
Simple code improvement

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx:
```diff
@@ -0,0 +1,167 @@
+  const task = tasks[index];
+  const src = imageField ? (task?.data?.[imageField] || "") : "";
+
+  const goToNext = () => {
+    if (index < tasks.length - 1) {
+      setCurrentTaskId(tasks[index + 1].id);
+    }
+  };
```
```suggestion
  const goToNext = useCallback(() => {
    if (index < tasks.length - 1) {
      setCurrentTaskId(tasks[index + 1].id);
    }
  }, [tasks]);
```

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx:
```diff
@@ -0,0 +1,167 @@
+    }
+  };
+
+  const goToPrev = () => {
+    if (index > 0) {
+      setCurrentTaskId(tasks[index - 1].id);
+    }
+  };
```
```suggestion
  const goToPrev = useCallback(() => {
    if (index > 0) {
      setCurrentTaskId(tasks[index - 1].id);
    }
  }, [tasks]);
```

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx:
```diff
@@ -0,0 +1,167 @@
+    }
+  };
+
+  const onSelect = () => view.toggleSelected(task.id);
```
```suggestion
  const onSelect = useCallback(() => view.toggleSelected(task?.id), [view.toggleSelected, task?.id]);
```

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx:
```diff
@@ -0,0 +1,167 @@
+
+  const onSelect = () => view.toggleSelected(task.id);
+
+  const onClose = () => setCurrentTaskId(null);
```
```suggestion
  const onClose = useCallback(() => setCurrentTaskId(null), []);
```

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx:
```diff
@@ -0,0 +1,167 @@
+
+    document.addEventListener("keydown", onKeyDown);
+    return () => document.removeEventListener("keydown", onKeyDown);
+  }, [goToNext, goToPrev]);
```
```suggestion
  }, [goToNext, goToPrev, onSelect, onClose]);
```

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridView.jsx:
```diff
@@ -56,11 +57,22 @@ const GridDataGroup = observer(({ type, value, field, row }) => {
 const GridCell = observer(({ view, selected, row, fields, onClick, ...props }) => {
+  const { setCurrentTaskId, imageField } = useContext(GridViewContext);
+
+  const handleBodyClick = (e) => {
```
I have seen more times that you have chosen not to wrap an inline function with useCallback than not, can you maybe explain the thinking? Generally speaking it is safer to default to always wrapping with useCallback to ensure there are no stale closures and that it is a proper dependency of any downstream components which are memoized, or as dependencies of other useCallback, useEffect or useMemo.

hlomzik replied:
the modal is just rendered only once for a task, so I didn't care, but you are right for sure, will update them in a sec

bmartel commented on web/libs/datamanager/src/components/MainView/GridView/GridView.jsx:
```diff
@@ -56,11 +57,22 @@ const GridDataGroup = observer(({ type, value, field, row }) => {
 const GridCell = observer(({ view, selected, row, fields, onClick, ...props }) => {
+  const { setCurrentTaskId, imageField } = useContext(GridViewContext);
+
+  const handleBodyClick = (e) => {
+    if (!isFF(FF_GRID_PREVIEW) || !imageField) return;
+    // @todo skip this interaction if there are no images in the task
+    e.stopPropagation();
+    setCurrentTaskId(row.id);
+  };
```
```suggestion
  const handleBodyClick = useCallback((e) => {
    if (!isFF(FF_GRID_PREVIEW) || !imageField) return;
    // @todo skip this interaction if there are no images in the task
    e.stopPropagation();
    setCurrentTaskId(row.id);
  }, [imageField, row.id]);
```

hlomzik committed:
Wrap methods in `useCallback()`

hlomzik committed:
Wrap one more function with `useCallback()`

Co-authored-by: bmartel <brandonmartel@gmail.com>

MihajloHoma commented:
/git merge 
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12277603253)
> Successfully merged:  create mode 100644 web/libs/frontend-test/src/helpers/utils/fixLSParams.ts

MihajloHoma committed:
Merge branch 'develop' into 'fb-leap-1424/grid-preview'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/12277603253

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12282247940)
> Successfully merged: Already up to date.

hlomzik committed:
Prevent Quick View from opening in a background by hotkey

Apparently `Shift+Up` opens up a Quick View even in a background

hlomzik committed:
Fix hotkeys interception for background hotkeys

MihajloHoma commented:
/git merge


> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12311351692)
> Successfully merged:  3 files changed, 83 insertions(+), 2 deletions(-)

MihajloHoma committed:
Merge branch 'develop' into 'fb-leap-1424/grid-preview'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/12311351692

hlomzik committed:
Properly handle view change

When we leave the Grid View by browser controls or by hotkeys we need to close the modal.
That also means that we don't need to intercept Quick View hotkey.
And `capture` was a huge hit for performance, because keydown handler was called million times
during handling the same event, leading to browser freeze.

hlomzik committed:
Fix linting

hlomzik committed:
Fix silly mistake

useEffect closing the modal should be global one-time one

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12355096988)
> Successfully merged:  12 files changed, 699 insertions(+), 40 deletions(-)

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1424/grid-preview'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/12355096988

hlomzik commented:
merged already by https://github.com/HumanSignal/label-studio/pull/6781
